### ISSUE TITLE
Serve Fails while Pipeline is working

### ISSUE THREAD

**Issue by aliozts** [microsoft/DeepSpeed-MII#266]
Hello, thank you for your efforts! I was trying to use the library for `mistral` deployment. 

```python
import mii
pipe = mii.pipeline("mistralai/Mistral-7B-v0.1")
response = pipe("DeepSpeed is", max_new_tokens=128)
print(response)
```

While I can run this code example with `deepspeed --num_gpus=2 main.py`

```python
import mii
client = mii.serve("mistralai/Mistral-7B-v0.1", tensor_parallel=2)
```

This fails with `deepspeed --num_gpus=2 main.py` (I have also tried not using `tensor_parallel` and using `deepspeed --num_gpus=2` only but regardless, it does not run with `deepspeed`)

I can run this with `python main.py` and I can connect with `mii.client` and generate but my GPU memory allocation is

<img width="624" alt="image" src="https://github.com/microsoft/DeepSpeed-MII/assets/36827959/3a864aeb-95de-41dd-a1fe-04c2566196e8">

Moreover, when I use `client.terminate_server()` only one of the GPU memory is cleaned

<img width="624" alt="image" src="https://github.com/microsoft/DeepSpeed-MII/assets/36827959/86c49382-44ed-43b7-8c4a-32bba3208b07">

I'm using 2 x L4 GPUs and working on GCP instance. I can provide any details if it will help. Thank you again

**Comment by mrwyattii** [comment]
Hi @aliozts When using the persistent deployment (i.e., `mii.serve()`), you should launch with `python` instead of `deepspeed`. You can still use DeepSpeed launcher, but you should make sure to use `deepspeed --num_gpus=1`. The reason for this is that `mii.serve("mistralai/Mistral-7B-v0.1", tensor_parallel=2)` uses the DeepSpeed launcher to stand up the GRPC server. When you do `deepspeed --num_gpus 2 main.py`, you are launching 2 servers, each trying to deploy across 2 GPUs.

The `client.terminate_server()` behavior is a bug and was patched in https://github.com/microsoft/DeepSpeed-MII/pull/262

**Comment by aliozts** [comment]
Hi @mrwyattii  thank you so much for the fix! I just installed from source and it's working. May I ask if it's normal for the model to take 21GBs of VRAM per GPU for a Llama 7B model? Does it allocate additional memory for requests or is there a limit to this?

**Comment by mrwyattii** [comment]
> May I ask if it's normal for the model to take 21GBs of VRAM per GPU for a Llama 7B model? Does it allocate additional memory for requests or is there a limit to this?

Yes, this is the default behavior. In general, we reserve as much GPU memory as we can for KV Cache. This behavior can be modified with the [Inference Engine Config](https://github.com/microsoft/DeepSpeed-MII/blob/7d37f27465e832d690b9e3e6d00156478e561f5d/mii/config.py#L53C21-L53C21), but we do not have it well documented yet (we're still working on improving things!).

**Comment by Traveller2001** [comment]
> Hi @aliozts When using the persistent deployment (i.e., `mii.serve()`), you should launch with `python` instead of `deepspeed`. You can still use DeepSpeed launcher, but you should make sure to use `deepspeed --num_gpus=1`. The reason for this is that `mii.serve("mistralai/Mistral-7B-v0.1", tensor_parallel=2)` uses the DeepSpeed launcher to stand up the GRPC server. When you do `deepspeed --num_gpus 2 main.py`, you are launching 2 servers, each trying to deploy across 2 GPUs.
> 
> The `client.terminate_server()` behavior is a bug and was patched in #262

hi,how can launch with python，this is my code.py
```
   import mii
   client = mii.serve("/data/hf_models/llama2-7b",tensor_parallel=4,device_map = [2,3,4,5])
   response = client.generate(["Deepspeed is"*100,], max_new_tokens=128)
   print(response)
```
and run `python code.py` ，but still launch many servers：
```
[2023-12-18 02:44:00,861] [INFO] [real_accelerator.py:161:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2023-12-18 02:44:01,637] [INFO] [server.py:107:_launch_server_process] msg_server launch: ['deepspeed', '-i', 'localhost:2,3,4,5', '--master_port', '29500', '--master_addr', 'localhost', '--no_ssh_check', '--no_local_rank', '--no_python', '/home/test_user/miniconda3/envs/infer/bin/python', '-m', 'mii.launch.multi_gpu_server', '--deployment-name', 'llama2-7b-mii-deployment', '--load-balancer-port', '50050', '--restful-gateway-port', '51080', '--restful-gateway-procs', '32', '--server-port', '50051', '--zmq-port', '25555', '--model-config', 'eyJtb2RlbF9uYW1lX29yX3BhdGgiOiAiL2RhdGEvaGZfbW9kZWxzL2xsYW1hMi03YiIsICJ0b2tlbml6ZXIiOiAiL2RhdGEvaGZfbW9kZWxzL2xsYW1hMi03YiIsICJ0YXNrIjogInRleHQtZ2VuZXJhdGlvbiIsICJ0ZW5zb3JfcGFyYWxsZWwiOiA0LCAiaW5mZXJlbmNlX2VuZ2luZV9jb25maWciOiB7InRlbnNvcl9wYXJhbGxlbCI6IHsidHBfc2l6ZSI6IDR9LCAic3RhdGVfbWFuYWdlciI6IHsibWF4X3RyYWNrZWRfc2VxdWVuY2VzIjogMjA0OCwgIm1heF9yYWdnZWRfYmF0Y2hfc2l6ZSI6IDc2OCwgIm1heF9yYWdnZWRfc2VxdWVuY2VfY291bnQiOiA1MTIsICJtYXhfY29udGV4dCI6IDgxOTIsICJtZW1vcnlfY29uZmlnIjogeyJtb2RlIjogInJlc2VydmUiLCAic2l6ZSI6IDEwMDAwMDAwMDB9LCAib2ZmbG9hZCI6IGZhbHNlfX0sICJ0b3JjaF9kaXN0X3BvcnQiOiAyOTUwMCwgInptcV9wb3J0X251bWJlciI6IDI1NTU1LCAicmVwbGljYV9udW0iOiAxLCAicmVwbGljYV9jb25maWdzIjogW3siaG9zdG5hbWUiOiAibG9jYWxob3N0IiwgInRlbnNvcl9wYXJhbGxlbF9wb3J0cyI6IFs1MDA1MSwgNTAwNTIsIDUwMDUzLCA1MDA1NF0sICJ0b3JjaF9kaXN0X3BvcnQiOiAyOTUwMCwgImdwdV9pbmRpY2VzIjogWzIsIDMsIDQsIDVdLCAiem1xX3BvcnQiOiAyNTU1NX1dLCAiZGV2aWNlX21hcCI6IHsibG9jYWxob3N0IjogW1syLCAzLCA0LCA1XV19LCAibWF4X2xlbmd0aCI6IG51bGwsICJhbGxfcmFua19vdXRwdXQiOiBmYWxzZSwgInN5bmNfZGVidWciOiBmYWxzZSwgInByb2ZpbGVfbW9kZWxfdGltZSI6IGZhbHNlfQ==']
[2023-12-18 02:44:01,637] [INFO] [server.py:107:_launch_server_process] msg_server launch: ['deepspeed', '-i', 'localhost:2,3,4,5', '--master_port', '29500', '--master_addr', 'localhost', '--no_ssh_check', '--no_local_rank', '--no_python', '/home/test_user/miniconda3/envs/infer/bin/python', '-m', 'mii.launch.multi_gpu_server', '--deployment-name', 'llama2-7b-mii-deployment', '--load-balancer-port', '50050', '--restful-gateway-port', '51080', '--restful-gateway-procs', '32', '--server-port', '50051', '--zmq-port', '25555', '--model-config', 'eyJtb2RlbF9uYW1lX29yX3BhdGgiOiAiL2RhdGEvaGZfbW9kZWxzL2xsYW1hMi03YiIsICJ0b2tlbml6ZXIiOiAiL2RhdGEvaGZfbW9kZWxzL2xsYW1hMi03YiIsICJ0YXNrIjogInRleHQtZ2VuZXJhdGlvbiIsICJ0ZW5zb3JfcGFyYWxsZWwiOiA0LCAiaW5mZXJlbmNlX2VuZ2luZV9jb25maWciOiB7InRlbnNvcl9wYXJhbGxlbCI6IHsidHBfc2l6ZSI6IDR9LCAic3RhdGVfbWFuYWdlciI6IHsibWF4X3RyYWNrZWRfc2VxdWVuY2VzIjogMjA0OCwgIm1heF9yYWdnZWRfYmF0Y2hfc2l6ZSI6IDc2OCwgIm1heF9yYWdnZWRfc2VxdWVuY2VfY291bnQiOiA1MTIsICJtYXhfY29udGV4dCI6IDgxOTIsICJtZW1vcnlfY29uZmlnIjogeyJtb2RlIjogInJlc2VydmUiLCAic2l6ZSI6IDEwMDAwMDAwMDB9LCAib2ZmbG9hZCI6IGZhbHNlfX0sICJ0b3JjaF9kaXN0X3BvcnQiOiAyOTUwMCwgInptcV9wb3J0X251bWJlciI6IDI1NTU1LCAicmVwbGljYV9udW0iOiAxLCAicmVwbGljYV9jb25maWdzIjogW3siaG9zdG5hbWUiOiAibG9jYWxob3N0IiwgInRlbnNvcl9wYXJhbGxlbF9wb3J0cyI6IFs1MDA1MSwgNTAwNTIsIDUwMDUzLCA1MDA1NF0sICJ0b3JjaF9kaXN0X3BvcnQiOiAyOTUwMCwgImdwdV9pbmRpY2VzIjogWzIsIDMsIDQsIDVdLCAiem1xX3BvcnQiOiAyNTU1NX1dLCAiZGV2aWNlX21hcCI6IHsibG9jYWxob3N0IjogW1syLCAzLCA0LCA1XV19LCAibWF4X2xlbmd0aCI6IG51bGwsICJhbGxfcmFua19vdXRwdXQiOiBmYWxzZSwgInN5bmNfZGVidWciOiBmYWxzZSwgInByb2ZpbGVfbW9kZWxfdGltZSI6IGZhbHNlfQ==']
[2023-12-18 02:44:01,656] [INFO] [server.py:107:_launch_server_process] msg_server launch: ['/home/test_user/miniconda3/envs/infer/bin/python', '-m', 'mii.launch.multi_gpu_server', '--deployment-name', 'llama2-7b-mii-deployment', '--load-balancer-port', '50050', '--restful-gateway-port', '51080', '--restful-gateway-procs', '32', '--load-balancer', '--model-config', 'eyJtb2RlbF9uYW1lX29yX3BhdGgiOiAiL2RhdGEvaGZfbW9kZWxzL2xsYW1hMi03YiIsICJ0b2tlbml6ZXIiOiAiL2RhdGEvaGZfbW9kZWxzL2xsYW1hMi03YiIsICJ0YXNrIjogInRleHQtZ2VuZXJhdGlvbiIsICJ0ZW5zb3JfcGFyYWxsZWwiOiA0LCAiaW5mZXJlbmNlX2VuZ2luZV9jb25maWciOiB7InRlbnNvcl9wYXJhbGxlbCI6IHsidHBfc2l6ZSI6IDR9LCAic3RhdGVfbWFuYWdlciI6IHsibWF4X3RyYWNrZWRfc2VxdWVuY2VzIjogMjA0OCwgIm1heF9yYWdnZWRfYmF0Y2hfc2l6ZSI6IDc2OCwgIm1heF9yYWdnZWRfc2VxdWVuY2VfY291bnQiOiA1MTIsICJtYXhfY29udGV4dCI6IDgxOTIsICJtZW1vcnlfY29uZmlnIjogeyJtb2RlIjogInJlc2VydmUiLCAic2l6ZSI6IDEwMDAwMDAwMDB9LCAib2ZmbG9hZCI6IGZhbHNlfX0sICJ0b3JjaF9kaXN0X3BvcnQiOiAyOTUwMCwgInptcV9wb3J0X251bWJlciI6IDI1NTU1LCAicmVwbGljYV9udW0iOiAxLCAicmVwbGljYV9jb25maWdzIjogW3siaG9zdG5hbWUiOiAibG9jYWxob3N0IiwgInRlbnNvcl9wYXJhbGxlbF9wb3J0cyI6IFs1MDA1MSwgNTAwNTIsIDUwMDUzLCA1MDA1NF0sICJ0b3JjaF9kaXN0X3BvcnQiOiAyOTUwMCwgImdwdV9pbmRpY2VzIjogWzIsIDMsIDQsIDVdLCAiem1xX3BvcnQiOiAyNTU1NX1dLCAiZGV2aWNlX21hcCI6IHsibG9jYWxob3N0IjogW1syLCAzLCA0LCA1XV19LCAibWF4X2xlbmd0aCI6IG51bGwsICJhbGxfcmFua19vdXRwdXQiOiBmYWxzZSwgInN5bmNfZGVidWciOiBmYWxzZSwgInByb2ZpbGVfbW9kZWxfdGltZSI6IGZhbHNlfQ==']
[2023-12-18 02:44:01,656] [INFO] [server.py:107:_launch_server_process] msg_server launch: ['/home/test_user/miniconda3/envs/infer/bin/python', '-m', 'mii.launch.multi_gpu_server', '--deployment-name', 'llama2-7b-mii-deployment', '--load-balancer-port', '50050', '--restful-gateway-port', '51080', '--restful-gateway-procs', '32', '--load-balancer', '--model-config', 'eyJtb2RlbF9uYW1lX29yX3BhdGgiOiAiL2RhdGEvaGZfbW9kZWxzL2xsYW1hMi03YiIsICJ0b2tlbml6ZXIiOiAiL2RhdGEvaGZfbW9kZWxzL2xsYW1hMi03YiIsICJ0YXNrIjogInRleHQtZ2VuZXJhdGlvbiIsICJ0ZW5zb3JfcGFyYWxsZWwiOiA0LCAiaW5mZXJlbmNlX2VuZ2luZV9jb25maWciOiB7InRlbnNvcl9wYXJhbGxlbCI6IHsidHBfc2l6ZSI6IDR9LCAic3RhdGVfbWFuYWdlciI6IHsibWF4X3RyYWNrZWRfc2VxdWVuY2VzIjogMjA0OCwgIm1heF9yYWdnZWRfYmF0Y2hfc2l6ZSI6IDc2OCwgIm1heF9yYWdnZWRfc2VxdWVuY2VfY291bnQiOiA1MTIsICJtYXhfY29udGV4dCI6IDgxOTIsICJtZW1vcnlfY29uZmlnIjogeyJtb2RlIjogInJlc2VydmUiLCAic2l6ZSI6IDEwMDAwMDAwMDB9LCAib2ZmbG9hZCI6IGZhbHNlfX0sICJ0b3JjaF9kaXN0X3BvcnQiOiAyOTUwMCwgInptcV9wb3J0X251bWJlciI6IDI1NTU1LCAicmVwbGljYV9udW0iOiAxLCAicmVwbGljYV9jb25maWdzIjogW3siaG9zdG5hbWUiOiAibG9jYWxob3N0IiwgInRlbnNvcl9wYXJhbGxlbF9wb3J0cyI6IFs1MDA1MSwgNTAwNTIsIDUwMDUzLCA1MDA1NF0sICJ0b3JjaF9kaXN0X3BvcnQiOiAyOTUwMCwgImdwdV9pbmRpY2VzIjogWzIsIDMsIDQsIDVdLCAiem1xX3BvcnQiOiAyNTU1NX1dLCAiZGV2aWNlX21hcCI6IHsibG9jYWxob3N0IjogW1syLCAzLCA0LCA1XV19LCAibWF4X2xlbmd0aCI6IG51bGwsICJhbGxfcmFua19vdXRwdXQiOiBmYWxzZSwgInN5bmNfZGVidWciOiBmYWxzZSwgInByb2ZpbGVfbW9kZWxfdGltZSI6IGZhbHNlfQ==']
[2023-12-18 02:44:03,649] [INFO] [real_accelerator.py:161:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2023-12-18 02:44:03,673] [INFO] [real_accelerator.py:161:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2023-12-18 02:44:04,623] [WARNING] [runner.py:202:fetch_hostfile] Unable to find hostfile, will proceed with training with local resources only.
Detected CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 but ignoring it because one or several of --include/--exclude/--num_gpus/--num_nodes cl args were used. If you want to use CUDA_VISIBLE_DEVICES don't pass any of these arguments to deepspeed.
[2023-12-18 02:44:04,623] [INFO] [runner.py:570:main] cmd = /home/test_user/miniconda3/envs/infer/bin/python -u -m deepspeed.launcher.launch --world_info=eyJsb2NhbGhvc3QiOiBbMiwgMywgNCwgNV19 --master_addr=127.0.0.1 --master_port=29500 --no_python --no_local_rank --enable_each_rank_log=None /home/test_user/miniconda3/envs/infer/bin/python -m mii.launch.multi_gpu_server --deployment-name llama2-7b-mii-deployment --load-balancer-port 50050 --restful-gateway-port 51080 --restful-gateway-procs 32 --server-port 50051 --zmq-port 25555 --model-config eyJtb2RlbF9uYW1lX29yX3BhdGgiOiAiL2RhdGEvaGZfbW9kZWxzL2xsYW1hMi03YiIsICJ0b2tlbml6ZXIiOiAiL2RhdGEvaGZfbW9kZWxzL2xsYW1hMi03YiIsICJ0YXNrIjogInRleHQtZ2VuZXJhdGlvbiIsICJ0ZW5zb3JfcGFyYWxsZWwiOiA0LCAiaW5mZXJlbmNlX2VuZ2luZV9jb25maWciOiB7InRlbnNvcl9wYXJhbGxlbCI6IHsidHBfc2l6ZSI6IDR9LCAic3RhdGVfbWFuYWdlciI6IHsibWF4X3RyYWNrZWRfc2VxdWVuY2VzIjogMjA0OCwgIm1heF9yYWdnZWRfYmF0Y2hfc2l6ZSI6IDc2OCwgIm1heF9yYWdnZWRfc2VxdWVuY2VfY291bnQiOiA1MTIsICJtYXhfY29udGV4dCI6IDgxOTIsICJtZW1vcnlfY29uZmlnIjogeyJtb2RlIjogInJlc2VydmUiLCAic2l6ZSI6IDEwMDAwMDAwMDB9LCAib2ZmbG9hZCI6IGZhbHNlfX0sICJ0b3JjaF9kaXN0X3BvcnQiOiAyOTUwMCwgInptcV9wb3J0X251bWJlciI6IDI1NTU1LCAicmVwbGljYV9udW0iOiAxLCAicmVwbGljYV9jb25maWdzIjogW3siaG9zdG5hbWUiOiAibG9jYWxob3N0IiwgInRlbnNvcl9wYXJhbGxlbF9wb3J0cyI6IFs1MDA1MSwgNTAwNTIsIDUwMDUzLCA1MDA1NF0sICJ0b3JjaF9kaXN0X3BvcnQiOiAyOTUwMCwgImdwdV9pbmRpY2VzIjogWzIsIDMsIDQsIDVdLCAiem1xX3BvcnQiOiAyNTU1NX1dLCAiZGV2aWNlX21hcCI6IHsibG9jYWxob3N0IjogW1syLCAzLCA0LCA1XV19LCAibWF4X2xlbmd0aCI6IG51bGwsICJhbGxfcmFua19vdXRwdXQiOiBmYWxzZSwgInN5bmNfZGVidWciOiBmYWxzZSwgInByb2ZpbGVfbW9kZWxfdGltZSI6IGZhbHNlfQ==
Starting load balancer on port: 50050
About to start server
Started
[2023-12-18 02:44:06,614] [INFO] [real_accelerator.py:161:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2023-12-18 02:44:06,675] [INFO] [server.py:65:_wait_until_server_is_live] waiting for server to start...
[2023-12-18 02:44:06,675] [INFO] [server.py:65:_wait_until_server_is_live] waiting for server to start...
[2023-12-18 02:44:07,502] [INFO] [launch.py:145:main] WORLD INFO DICT: {'localhost': [2, 3, 4, 5]}
[2023-12-18 02:44:07,502] [INFO] [launch.py:151:main] nnodes=1, num_local_procs=4, node_rank=0
[2023-12-18 02:44:07,502] [INFO] [launch.py:162:main] global_rank_mapping=defaultdict(<class 'list'>, {'localhost': [0, 1, 2, 3]})
[2023-12-18 02:44:07,502] [INFO] [launch.py:163:main] dist_world_size=4
[2023-12-18 02:44:07,502] [INFO] [launch.py:165:main] Setting CUDA_VISIBLE_DEVICES=2,3,4,5
[2023-12-18 02:44:09,651] [INFO] [real_accelerator.py:161:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2023-12-18 02:44:09,658] [INFO] [real_accelerator.py:161:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2023-12-18 02:44:09,687] [INFO] [real_accelerator.py:161:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2023-12-18 02:44:09,793] [INFO] [real_accelerator.py:161:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2023-12-18 02:44:10,786] [INFO] [comm.py:637:init_distributed] cdb=None
[2023-12-18 02:44:10,786] [INFO] [comm.py:668:init_distributed] Initializing TorchBackend in DeepSpeed with backend nccl
[W socket.cpp:436] [c10d] The server socket has failed to bind to [::]:29500 (errno: 98 - Address already in use).
[W socket.cpp:436] [c10d] The server socket has failed to bind to 0.0.0.0:29500 (errno: 98 - Address already in use).
[E socket.cpp:472] [c10d] The server socket has failed to listen on any local network address.
```
why it still launch with deepspeed
